app/grandchallenge/cases/image_builders/dicom_4dct.py

```diff
@@ -0,0 +1,191 @@
+import tempfile
+from pathlib import Path
+
+import SimpleITK
+import numpy as np
+import pydicom
+
+from grandchallenge.cases.image_builders import ImageBuilderResult
+from grandchallenge.cases.image_builders.utils import convert_itk_to_internal
+
+NUMPY_IMAGE_TYPES = {
+    "character": SimpleITK.sitkUInt8,
+    "uint8": SimpleITK.sitkUInt8,
+    "uint16": SimpleITK.sitkUInt16,
+    "uint32": SimpleITK.sitkUInt32,
+    "uint64": SimpleITK.sitkUInt64,
+    "int8": SimpleITK.sitkInt8,
+    "int16": SimpleITK.sitkInt16,
+    "int32": SimpleITK.sitkInt32,
+    "int64": SimpleITK.sitkInt64,
+    "float32": SimpleITK.sitkFloat32,
+    "float64": SimpleITK.sitkFloat64,
+}
+
+
+def pixel_data_reached(tag, vr, length):
+    if pydicom.datadict.keyword_for_tag(tag) == "PixelData":
+        return True
+    return False
+
+
+def _get_headers(path):
+    """
+    Gets all headers from dicom files found in path.
+
+    Parameters
+    ----------
+    path: Path
+        Path to a directory that contains all images that were uploaded during
+        an upload session.
+
+    Returns
+    -------
+    Sorted headers for all dicom image files found within path.
+
+    Raises
+    ------
+    ValueError
+        If not all files are dicom files or if the study id varies between
+        files.
+    """
+    headers = []
+    study_id = None
+    for file in path.iterdir():
+        if not file.is_file():
+            continue
+        with file.open("rb") as f:
+            try:
+                ds = pydicom.filereader.read_partial(
+                    f, stop_when=pixel_data_reached
+                )
+                headers.append({"file": str(file), "data": ds})
+                if study_id and ds.StudyID != study_id:
+                    raise ValueError("Study ID is inconsistent across files")
+            except Exception:
+                raise ValueError("Invalid dicom file passed.")
+    headers.sort(key=lambda x: x["data"].InStackPositionNumber)
+    return headers
+
+
+def _validate_dicom_files(path):
+    """
+    Gets the headers for all dicom files on path and validates them.
+
+    Parameters
+    ----------
+    path: Path
+        Path to a directory that contains all images that were uploaded during
+        an upload session.
+
+    Returns
+    -------
+    A tuple of
+     - Headers for all dicom image files found within path
+     - Number of time points
+     - Number of slices per time point
+
+    Raises
+    ------
+    ValueError
+        If the number of slices varies between time points.
+    """
+    headers = _get_headers(path)
+    if not headers:
+        raise ValueError("No dicom headers could be parsed")
+    n_time = headers[-1]["data"].TemporalPositionIndex
+    if len(headers) % n_time > 0:
+        raise ValueError("Number of slices per time point varies")
+    n_slices = len(headers) // n_time
+    return headers, n_time, n_slices
```

Comment:
Another nitpick: I generally prefer returning NamedTuples rather than tuples, see https://docs.quantifiedcode.com/python-anti-patterns/readability/not_using_named_tuples_when_returning_more_than_one_value.html